File: OsmAnd/src/net/osmand/plus/wikivoyage/data/TravelArticle.java
package net.osmand.plus.wikivoyage.data;

import net.osmand.gpx.GPXFile;
import net.osmand.gpx.GPXTrackAnalysis;
import static net.osmand.gpx.GPXUtilities.WptPt;
import static net.osmand.osm.MapPoiTypes.ROUTE_ARTICLE_POINT;
import static net.osmand.util.Algorithms.capitalizeFirstLetter;

import android.os.Parcel;
import android.os.Parcelable;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.Size;

import net.osmand.IndexConstants;
import net.osmand.PlatformUtil;
import net.osmand.binary.BinaryMapIndexReader;
import net.osmand.data.Amenity;
import net.osmand.osm.PoiCategory;
import net.osmand.plus.OsmandApplication;
import net.osmand.plus.card.color.palette.main.data.DefaultColors;
import net.osmand.util.Algorithms;

import org.apache.commons.codec.binary.Hex;
import org.apache.commons.codec.digest.DigestUtils;
import org.apache.commons.logging.Log;

import java.io.File;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;

public class TravelArticle {

	private static final Log LOG = PlatformUtil.getLog(TravelArticle.class);
	private static final String IMAGE_ROOT_URL = "https://upload.wikimedia.org/wikipedia/commons/";
	private static final String THUMB_PREFIX = "320px-";
	private static final String REGULAR_PREFIX = "1280px-";//1280, 1024, 800


	File file;
	String title;
	String content;
	String isPartOf;
	String isParentOf = "";
	double lat = Double.NaN;
	double lon = Double.NaN;
	String imageTitle;
	GPXFile gpxFile;
	String routeId;
	int routeRadius = -1;
	public String ref;
	String routeSource = "";
	long originalId;
	String lang;
	String contentsJson;
	String aggregatedPartOf;
	String description;

	long lastModified;
	boolean gpxFileReading;
	boolean gpxFileRead;

	@NonNull
	public TravelArticleIdentifier generateIdentifier() {
		return new TravelArticleIdentifier(this);
	}

	@NonNull
	public static String getTravelBook(@NonNull OsmandApplication app, @NonNull File file) {
		return file.getPath().replace(app.getAppPath(IndexConstants.WIKIVOYAGE_INDEX_DIR).getPath() + "/", "");
	}

	@Nullable
	public String getTravelBook(@NonNull OsmandApplication app) {
		return file != null ? getTravelBook(app, file) : null;
	}

	public File getFile() {
		return file;
	}

	public long getLastModified() {
		if (lastModified > 0) {
			return lastModified;
		}
		return file != null ? file.lastModified() : 0;
	}

	public String getTitle() {
		return title;
	}

	public String getDescription() {
		return description;
	}

	public String getContent() {
		return content;
	}

	public String getIsPartOf() {
		return isPartOf;
	}

	public double getLat() {
		return lat;
	}

	public double getLon() {
		return lon;
	}

	public String getImageTitle() {
		return imageTitle;
	}

	public GPXFile getGpxFile() {
		return gpxFile;
	}

	public String getRouteId() {
		return routeId;
	}

	public String getRouteSource() {
		return routeSource;
	}

	public long getOriginalId() {
		return originalId;
	}

	public String getLang() {
		return lang;
	}

	public String getContentsJson() {
		return contentsJson;
	}

	public String getAggregatedPartOf() {
		return aggregatedPartOf;
	}

	@Nullable
	public String getGeoDescription() {
		if (TextUtils.isEmpty(aggregatedPartOf)) {
			return null;
		}

		String[] parts = aggregatedPartOf.split(",");
		if (parts.length > 0) {
			StringBuilder res = new StringBuilder();
			res.append(parts[parts.length - 1]);
			if (parts.length > 1) {
				res.append(" \u2022 ").append(parts[0]);
			}
			return res.toString();
		}

		return null;
	}

	@NonNull
	public static String getImageUrl(@NonNull String imageTitle, boolean thumbnail) {
		imageTitle = imageTitle.replace(" ", "_");
		try {
			imageTitle = URLDecoder.decode(imageTitle, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			System.err.println(e.getMessage());
		}
		String[] hash = getHash(imageTitle);
		try {
			imageTitle = URLEncoder.encode(imageTitle, "UTF-8");
		} catch (UnsupportedEncodingException e) {
			System.err.println(e.getMessage());
		}
		String prefix = thumbnail ? THUMB_PREFIX : REGULAR_PREFIX;
		String suffix = imageTitle.endsWith(".svg") ? ".png" : "";
		return IMAGE_ROOT_URL + "thumb/" + hash[0] + "/" + hash[1] + "/" + imageTitle + "/" + prefix + imageTitle + suffix;
	}

	@NonNull
	public String getPointFilterString(){
		return ROUTE_ARTICLE_POINT;
	}

	@NonNull
	public WptPt createWptPt(@NonNull Amenity amenity, @Nullable String lang) {
		WptPt wptPt = new WptPt();
		wptPt.name = amenity.getName();
		wptPt.lat = amenity.getLocation().getLatitude();
		wptPt.lon = amenity.getLocation().getLongitude();
		wptPt.desc = amenity.getDescription(lang);
		wptPt.link = amenity.getSite();
		String colorId = amenity.getColor();
		if (colorId != null) {
			wptPt.setColor(DefaultColors.valueOf(colorId));
		}
		String iconName = amenity.getGpxIcon();
		if (iconName != null) {
			wptPt.setIconName(iconName);
		}
		String category = amenity.getTagSuffix("category_");
		if (category != null) {
			wptPt.category = capitalizeFirstLetter(category);
		}
		return wptPt;
	}

	@Nullable
	public GPXTrackAnalysis getAnalysis() {
		return null;
	}

	@NonNull
	public BinaryMapIndexReader.SearchPoiTypeFilter getSearchFilter(String filterSubcategory) {
		return new BinaryMapIndexReader.SearchPoiTypeFilter() {
			@Override
			public boolean accept(PoiCategory type, String subcategory) {
				return subcategory.equals(filterSubcategory);
			}

			@Override
			public boolean isEmpty() {
				return false;
			}
		};
	}

	@Size(2)
	@NonNull
	private static String[] getHash(@NonNull String s) {
		String md5 = new String(Hex.encodeHex(DigestUtils.md5(s)));
		return new String[]{md5.substring(0, 1), md5.substring(0, 2)};
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (o == null || getClass() != o.getClass()) {
			return false;
		}
		TravelArticle that = (TravelArticle) o;
		return TravelArticleIdentifier.areLatLonEqual(that.lat, that.lon, lat, lon) &&
				Algorithms.objectEquals(file, that.file) &&
				Algorithms.stringsEqual(routeId, that.routeId) &&
				Algorithms.stringsEqual(routeSource, that.routeSource);
	}

	@Override
	public int hashCode() {
		return Algorithms.hash(file, lat, lon, routeId, routeSource);
	}

	public static class TravelArticleIdentifier implements Parcelable {
		@Nullable File file;
		double lat;
		double lon;
		@Nullable String title;
		@Nullable String routeId;
		@Nullable String routeSource;

		public static final Creator<TravelArticleIdentifier> CREATOR = new Creator<TravelArticleIdentifier>() {
			@Override
			public TravelArticleIdentifier createFromParcel(Parcel in) {
				return new TravelArticleIdentifier(in);
			}

			@Override
			public TravelArticleIdentifier[] newArray(int size) {
				return new TravelArticleIdentifier[size];
			}
		};

		private TravelArticleIdentifier(@NonNull Parcel in) {
			readFromParcel(in);
		}

		private TravelArticleIdentifier(@NonNull TravelArticle article) {
			file = article.file;
			lat = article.lat;
			lon = article.lon;
			title = article.title;
			routeId = article.routeId;
			routeSource = article.routeSource;
		}

		@Override
		public void writeToParcel(Parcel out, int flags) {
			out.writeDouble(lat);
			out.writeDouble(lon);
			out.writeString(title);
			out.writeString(routeId);
			out.writeString(routeSource);
			out.writeString(file != null ? file.getAbsolutePath() : null);
		}

		private void readFromParcel(Parcel in) {
			lat = in.readDouble();
			lon = in.readDouble();
			title = in.readString();
			routeId = in.readString();
			routeSource = in.readString();
			String filePath = in.readString();
			if (!Algorithms.isEmpty(filePath)) {
				file = new File(filePath);
			}
		}

		@Override
		public int describeContents() {
			return 0;
		}

		@Override
		public boolean equals(Object o) {
			if (this == o) {
				return true;
			}
			if (o == null || getClass() != o.getClass()) {
				return false;
			}
			TravelArticleIdentifier that = (TravelArticleIdentifier) o;
			return areLatLonEqual(that.lat, that.lon, lat, lon) &&
					Algorithms.objectEquals(file, that.file) &&
					Algorithms.stringsEqual(routeId, that.routeId) &&
					Algorithms.stringsEqual(routeSource, that.routeSource);
		}

		@Override
		public int hashCode() {
			return Algorithms.hash(file, lat, lon, routeId, routeSource);
		}

		public static boolean areLatLonEqual(double lat1, double lon1, double lat2, double lon2) {
			boolean latEqual = (Double.isNaN(lat1) && Double.isNaN(lat2)) || Math.abs(lat1 - lat2) < 0.00001;
			boolean lonEqual = (Double.isNaN(lon1) && Double.isNaN(lon2)) || Math.abs(lon1 - lon2) < 0.00001;
			return latEqual && lonEqual;
		}
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/travel/TravelArticle.kt
package net.osmand.shared.travel

import net.osmand.shared.binary.SearchPoiTypeFilter
import net.osmand.shared.data.Amenity
import net.osmand.shared.data.KQuadRect
import net.osmand.shared.gpx.GpxFile
import net.osmand.shared.gpx.GpxTrackAnalysis
import net.osmand.shared.gpx.primitives.Link
import net.osmand.shared.gpx.primitives.WptPt
import net.osmand.shared.io.KFile
import net.osmand.shared.osm.MapPoiTypes
import net.osmand.shared.osm.PoiCategory
import net.osmand.shared.util.KAlgorithms
import net.osmand.shared.util.KMapUtils
import net.osmand.shared.util.LoggerFactory
import net.osmand.shared.util.UrlEncoder
import okio.ByteString.Companion.encodeUtf8
import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/**
 * One wikivoyage article, or one track stored in an obf file - [TravelGpx] is the second kind.
 * It carries what the poi section said about it and, once it has been built, the gpx itself.
 *
 * A copy of `TravelArticle` in the android app, which stays there; this copy is for iOS. Left out:
 * `Parcelable`, which is how android passes the identifier between screens, and `getTravelBook`,
 * which turns a path into a display name through the app's directory layout.
 */
open class TravelArticle {

	@JvmField
	var file: KFile? = null

	@JvmField
	var title: String? = null

	@JvmField
	var content: String? = null

	@JvmField
	var isPartOf: String? = null

	@JvmField
	var isParentOf: String = ""

	@JvmField
	var lat: Double = Double.NaN

	@JvmField
	var lon: Double = Double.NaN

	@JvmField
	var imageTitle: String? = null

	@JvmField
	var gpxFile: GpxFile? = null

	@JvmField
	var routeId: String? = null

	@JvmField
	var ref: String? = null

	@JvmField
	var routeSource: String = ""

	@JvmField
	var originalId: Long = 0

	@JvmField
	var lang: String? = null

	@JvmField
	var contentsJson: String? = null

	@JvmField
	var aggregatedPartOf: String? = null

	@JvmField
	var description: String? = null

	@JvmField
	var lastModified: Long = 0

	@JvmField
	var gpxFileReading: Boolean = false

	@JvmField
	var gpxFileRead: Boolean = false

	@JvmField
	var routeRadius: Int = -1

	private var bbox31: KQuadRect? = null

	/**
	 * The box the route covers, from the shortlink tiles the indexer wrote for it. It is what tells
	 * the reader which files to open before anything of the route has been read.
	 */
	fun initShortLinkTiles(shortLinkTiles: String) {
		val box = KQuadRect()
		this.bbox31 = box
		for (shortLink in shortLinkTiles.split(",")) {
			val bbox = KMapUtils.decodeShortLinkToQuadRect(shortLink)
			val left = KMapUtils.get31TileNumberX(bbox.left)
			val top = KMapUtils.get31TileNumberY(bbox.top)
			val right = KMapUtils.get31TileNumberX(bbox.right)
			val bottom = KMapUtils.get31TileNumberY(bbox.bottom)
			box.expand(left.toDouble(), top.toDouble(), right.toDouble(), bottom.toDouble())
		}
	}

	fun getBbox31(): KQuadRect? = bbox31

	fun hasBbox31(): Boolean = bbox31?.hasInitialState() == false

	fun generateIdentifier(): TravelArticleIdentifier = TravelArticleIdentifier(this)

	fun getFile(): KFile? = file

	fun getLastModified(): Long {
		if (lastModified > 0) {
			return lastModified
		}
		return file?.lastModified() ?: 0
	}

	fun getTitle(): String? = title

	fun getDescription(): String? = description

	fun getContent(): String? = content

	fun getIsPartOf(): String? = isPartOf

	fun getLat(): Double = lat

	fun getLon(): Double = lon

	fun getImageTitle(): String? = imageTitle

	fun getGpxFile(): GpxFile? = gpxFile

	fun getRouteId(): String? = routeId

	fun hasOsmRouteId(): Boolean {
		val routeId = getRouteId()
		return routeId != null &&
				(routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX_LEGACY) ||
						routeId.startsWith(Amenity.ROUTE_ID_OSM_PREFIX))
	}

	/** What the article is saved as; never empty, never a path. */
	fun getGpxFileName(): String {
		val gpxFileName = if (!KAlgorithms.isEmpty(title)) title else routeId
		if (gpxFileName != null) {
			return KAlgorithms.sanitizeFileName(gpxFileName)
		}
		log.error("Empty travel article in ${this.file}")
		return "Travel Article File"
	}

	fun getRouteSource(): String = routeSource

	fun getOriginalId(): Long = originalId

	fun getLang(): String? = lang

	fun getContentsJson(): String? = contentsJson

	fun getAggregatedPartOf(): String? = aggregatedPartOf

	/** The innermost and outermost regions the article belongs to, for the subtitle of a card. */
	fun getGeoDescription(): String? {
		val aggregatedPartOf = this.aggregatedPartOf
		if (KAlgorithms.isEmpty(aggregatedPartOf)) {
			return null
		}
		val parts = aggregatedPartOf!!.split(",")
		if (parts.isNotEmpty()) {
			val res = StringBuilder()
			res.append(parts[parts.size - 1])
			if (parts.size > 1) {
				res.append(" • ").append(parts[0])
			}
			return res.toString()
		}
		return null
	}

	open fun getPointFilterString(): String = MapPoiTypes.ROUTE_ARTICLE_POINT

	open fun getMainFilterString(): String = MapPoiTypes.ROUTE_ARTICLE

	/** One point of a wikivoyage article, with the tags that article points are written with. */
	open fun createWptPt(amenity: Amenity, lang: String?): WptPt {
		val wptPt = WptPt()
		wptPt.name = (amenity.getName())
		wptPt.lat = (amenity.getLocation()!!.latitude)
		wptPt.lon = (amenity.getLocation()!!.longitude)
		wptPt.desc = (amenity.getDescription(lang))
		wptPt.link = (Link(amenity.getSite() ?: ""))
		val colorId = amenity.getColor()
		if (colorId != null) {
			wptPt.setColor(defaultColor(colorId))
		}
		val iconName = amenity.getGpxIcon()
		if (iconName != null) {
			wptPt.setIconName(iconName)
		}
		val category = amenity.getTagSuffix("category_")
		if (category != null) {
			wptPt.category = (KAlgorithms.capitalizeFirstLetter(category))
		}
		for (key in amenity.getAdditionalInfoKeys()) {
			if (!WikivoyageOSMTags.contains(key)) {
				continue
			}
			val amenityAdditionalInfo = amenity.getAdditionalInfo(key)
			if (amenityAdditionalInfo != null) {
				wptPt.getExtensionsToWrite()[key] = amenityAdditionalInfo
			}
		}
		return wptPt
	}

	open fun getAnalysis(): GpxTrackAnalysis? = null

	/** Accepts one subtype, which is how a search is narrowed to the points of one article. */
	fun getSearchFilter(filterSubcategory: String): SearchPoiTypeFilter {
		return object : SearchPoiTypeFilter {
			override fun accept(type: PoiCategory?, subcategory: String): Boolean =
				subcategory == filterSubcategory

			override fun isEmpty(): Boolean = false
		}
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) {
			return true
		}
		if (other == null || this::class != other::class) {
			return false
		}
		other as TravelArticle
		return TravelArticleIdentifier.areLatLonEqual(other.lat, other.lon, lat, lon) &&
				file == other.file &&
				KAlgorithms.stringsEqual(routeId, other.routeId) &&
				KAlgorithms.stringsEqual(routeSource, other.routeSource)
	}

	override fun hashCode(): Int = KAlgorithms.hash(file, lat, lon, routeId, routeSource)

	companion object {
		private val log = LoggerFactory.getLogger("TravelArticle")

		private const val IMAGE_ROOT_URL = "https://upload.wikimedia.org/wikipedia/commons/"
		private const val THUMB_PREFIX = "330px-"
		private const val REGULAR_PREFIX = "1280px-" // 1280, 1024, 800

		const val TRAVEL_GPX_DEFAULT_SEARCH_RADIUS: Int = 50 * 1000

		/**
		 * Where wikimedia keeps the picture of that name: the first one and two characters of the
		 * md5 of the title are the two directories it lives under.
		 */
		@JvmStatic
		fun getImageUrl(imageTitle: String, thumbnail: Boolean): String {
			var title = imageTitle.replace(" ", "_")
			title = UrlEncoder.decode(title)
			val hash = getHash(title)
			title = UrlEncoder.encode(title)
			val prefix = if (thumbnail) THUMB_PREFIX else REGULAR_PREFIX
			val suffix = if (title.endsWith(".svg")) ".png" else ""
			return IMAGE_ROOT_URL + "thumb/" + hash[0] + "/" + hash[1] + "/" + title + "/" +
					prefix + title + suffix
		}

		private fun getHash(s: String): Array<String> {
			val md5 = s.encodeUtf8().md5().hex()
			return arrayOf(md5.substring(0, 1), md5.substring(0, 2))
		}

		/**
		 * The colour of that name in OsmAnd's own palette, or 0. A copy of `DefaultColors` in the
		 * android app, which keeps the names for the interface as well; these values are written
		 * into gpx files, so they are frozen.
		 */
		internal fun defaultColor(id: String): Int {
			return when (id) {
				"darkyellow" -> 0xffeecc22.toInt()
				"red" -> 0xffd00d0d.toInt()
				"orange" -> 0xffff5020.toInt()
				"yellow" -> 0xffeeee10.toInt()
				"lightgreen" -> 0xff88e030.toInt()
				"green" -> 0xff00842b.toInt()
				"lightblue" -> 0xff10c0f0.toInt()
				"blue" -> 0xff1010a0.toInt()
				"purple" -> 0xffa71de1.toInt()
				"pink" -> 0xffe044bb.toInt()
				"brown" -> 0xff8e2512.toInt()
				"black" -> 0xff000001.toInt()
				else -> 0
			}
		}
	}
}
